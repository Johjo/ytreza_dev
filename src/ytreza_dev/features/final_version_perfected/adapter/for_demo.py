from expression import Nothing
from pyqure import Key

from ytreza_dev.features.final_version_perfected.port.external_todolist import ExternalTodolistPort
from ytreza_dev.features.final_version_perfected.port.task_information_repository import TaskInformationRepositoryPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.port.task_reader import TaskFvpReaderPort
from ytreza_dev.features.final_version_perfected.port.task_repository import FvpRepositoryPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, ExternalTask


class TaskInMemory:
    def __init__(self) -> None:
        self._tasks: list[TaskBase] = []

    def all_tasks(self) -> list[TaskBase]:
        return self._tasks

    def save(self, tasks: list[TaskBase]) -> None:
        self._tasks = tasks


class TaskFvpReaderForDemo(TaskFvpReaderPort):
    def __init__(self, memory: TaskInMemory):
        self._memory = memory

    def all_active_tasks(self) -> list[TaskBase]:
        return self._memory.all_tasks()


class FvpRepositoryForDemo(FvpRepositoryPort):
    def all_tasks(self) -> list[TaskBase]:
        return self._memory.all_tasks()

    def __init__(self, memory: TaskInMemory):
        self._memory = memory

    def save(self, tasks: list[TaskBase]) -> None:
        self._memory.save(tasks)


TASK_IN_MEMORY_KEY = Key("task_in_memory", TaskInMemory)


class ExternalTodolistForDemo(ExternalTodolistPort):
    def close_task(self, url: str, task_id: str) -> None:
        pass


class TaskInformationRepositoryForDemo(TaskInformationRepositoryPort):
    def __init__(self):
        self._tasks: dict[str, TaskInformation] = {}

    def by_key(self, key: str) -> TaskInformation:
        return self._tasks[key]

    def save(self, task: list[TaskInformation]) -> None:
        for task in task:
            self._tasks[task.key] = task
