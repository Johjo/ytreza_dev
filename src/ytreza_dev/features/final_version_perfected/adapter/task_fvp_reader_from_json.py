import json
from pathlib import Path

from ytreza_dev.features.final_version_perfected.port.task_reader import TaskFvpReaderPort
from ytreza_dev.features.final_version_perfected.types import TaskBase, TaskNew, TaskNext, TaskLater, TaskNever, Project


class TaskFvpReaderFromJson(TaskFvpReaderPort):
    def __init__(self, json_path: Path) -> None:
        self._json_path = json_path

    def all_active_tasks(self) -> list[TaskBase]:
        tasks = json.loads(self._json_path.read_text(encoding="utf-8"))
        return [self._to_task(task) for task in tasks]

    def _to_task(self, task: dict[str, str]) -> TaskBase:
        match task["status"]:
            case "new":
                return TaskNew(title=task["title"], url=task["url"], id=task["id"])
            case "next":
                return TaskNext(title=task["title"], url=task["url"], id=task["id"])
            case "later":
                return TaskLater(title=task["title"], url=task["url"], id=task["id"])
            case "never":
                return TaskNever(title=task["title"], url=task["url"], id=task["id"])
        raise ValueError(f"Unknown status: {task['status']}")

    @staticmethod
    def _to_project(task) -> Project:
        return Project(key=task["project"]["key"], name=task["project"]["name"])

