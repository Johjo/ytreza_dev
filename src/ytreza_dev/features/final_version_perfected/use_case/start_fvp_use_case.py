from expression import Nothing

from ytreza_dev.features.final_version_perfected.port.task_information_repository import TaskInformationRepositoryPort, \
    TaskInformation
from ytreza_dev.features.final_version_perfected.port.task_repository import FvpRepositoryPort
from ytreza_dev.features.final_version_perfected.port.todolist_reader import TodolistReaderPort
from ytreza_dev.features.final_version_perfected.types import TaskNew, ExternalTask, ExternalProject, Project


class StartFvpUseCase:
    def __init__(self, todolist_reader: TodolistReaderPort, fvp_repository: FvpRepositoryPort, task_information_repository: TaskInformationRepositoryPort) -> None:
        self._task_information_repository = task_information_repository
        self._todolist_reader = todolist_reader
        self._fvp_repository = fvp_repository

    def execute(self) -> None:

        all_external_tasks : list[ExternalTask] = self._todolist_reader.all_tasks()
        self._fvp_repository.save([self._to_task_new(task) for task in all_external_tasks])
        self._task_information_repository.save([self._to_task_information(task) for task in all_external_tasks])

    @staticmethod
    def _to_task_information(external_task: ExternalTask) -> TaskInformation:
        return TaskInformation(
            key=external_task.id,
            title=external_task.name,
            project=Project(key=external_task.project.key, name=external_task.project.name),
            url=external_task.url,
            due_date=Nothing
        )

    @staticmethod
    def _to_task_new(task: ExternalTask) -> TaskNew :
        return TaskNew(id=task.id)

    @staticmethod
    def _to_project(project: ExternalProject) -> Project:
        return Project(key=project.key, name=project.name)
