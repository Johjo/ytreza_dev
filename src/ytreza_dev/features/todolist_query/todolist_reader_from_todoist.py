import os
import requests
from dotenv import load_dotenv

from ytreza_dev.features.todolist_query.todolist_query import TodolistReaderPort, Task

load_dotenv()


class TodolistReaderFromTodoist(TodolistReaderPort):
    def __init__(self):
        self.api_token = os.getenv("TODOIST_API_TOKEN")
        if not self.api_token:
            raise ValueError("TODOIST_API_TOKEN environment variable not set.")
        self.api_url = "https://api.todoist.com/rest/v2"

    def all_tasks(self) -> list[Task]:
        headers = {
            "Authorization": f"Bearer {self.api_token}",
        }
        response = requests.get(f"{self.api_url}/tasks", headers=headers)
        response.raise_for_status()  # Raise an exception for bad status codes
        tasks_data = response.json()
        return [Task(name=task_data["content"], url=task_data["url"]) for task_data in tasks_data]

