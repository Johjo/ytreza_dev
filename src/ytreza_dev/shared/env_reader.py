import os
from abc import ABC


class EnvReaderPort(ABC):
    def read(self, key) -> str:
        value = os.getenv(key)
        if value is None:
            raise KeyError(f"env key {key} is not defined")
        return value


class EnvReaderFromEnv(EnvReaderPort):
    pass
