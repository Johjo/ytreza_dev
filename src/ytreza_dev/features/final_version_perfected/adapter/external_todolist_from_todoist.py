from ytreza_dev.features.final_version_perfected.port.external_todolist import ExternalTodolistPort
from ytreza_dev.shared.env_reader import EnvReaderPort
from ytreza_dev.shared.todoist.todoist_api import TodoistAPI


class ExternalTodolistFromTodoist(ExternalTodolistPort):
    def __init__(self, env_reader: EnvReaderPort) -> None:
        self._todoist = TodoistAPI(env_reader.read("TODOIST_API_TOKEN"))

    def close_task(self, url: str, task_id: str) -> None:
        print(f"close task {task_id} on {url}")
        self._todoist.close_task(task_id=task_id)
