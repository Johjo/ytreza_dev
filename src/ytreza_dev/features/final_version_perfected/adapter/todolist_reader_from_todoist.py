from toml.encoder import TomlPreserveInlineDictEncoder

from ytreza_dev.features.final_version_perfected.port.todolist_reader import TodolistReaderPort
from ytreza_dev.features.final_version_perfected.types import ExternalTask, ExternalProject
from ytreza_dev.shared.env_reader import EnvReaderPort
from ytreza_dev.shared.todoist.todoist_api import TodoistAPI, TodoistTask, TodoistProject


class TodolistReaderFromTodoist(TodolistReaderPort):
    def __init__(self, env_reader: EnvReaderPort):
        self._env_reader = env_reader

    def all_tasks(self) -> list[ExternalTask]:
        todoist = TodoistAPI(self._env_reader.read("TODOIST_API_TOKEN"))
        tasks = todoist.get_all_tasks()
        return [self._to_external_task(task) for task in tasks]

    def _to_external_task(self, task: TodoistTask):
        return ExternalTask(name=task.name, url=task.url, id=task.id, project=self._to_external_project(task.project))

    def _to_external_project(self, project: TodoistProject) -> ExternalProject:
        return ExternalProject(name=project.name, key=project.id)
