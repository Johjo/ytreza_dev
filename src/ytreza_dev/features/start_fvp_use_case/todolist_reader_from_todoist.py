from ytreza_dev.features.start_fvp_use_case.use_case import TodolistReader, OldTask, ExternalTask
from ytreza_dev.shared.env_reader import EnvReaderPort, EnvReaderFromEnv
from ytreza_dev.shared.todoist.todoist_api import TodoistAPI


class TodolistReaderFromTodoist(TodolistReader):
    def __init__(self, env_reader: EnvReaderPort):
        self._env_reader = env_reader

    def all_tasks(self) -> list[ExternalTask]:
        todoist = TodoistAPI(self._env_reader.read("TODOIST_API_TOKEN"))
        tasks = todoist.get_all_tasks()
        return [ExternalTask(name=task.name, url=task.url) for task in tasks]
