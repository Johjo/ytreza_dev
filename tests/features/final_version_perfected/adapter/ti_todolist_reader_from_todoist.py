import os

from approvaltests import verify # type: ignore
from dotenv import load_dotenv

from ytreza_dev.features.final_version_perfected.adapter.todolist_reader_from_todoist import TodolistReaderFromTodoist
from ytreza_dev.shared.env_reader import EnvReaderFromEnv


def test_all_tasks() -> None:
    todoist_api = TodolistReaderFromTodoist(EnvReaderFromEnv(env_path=".env.test"))
    tasks = todoist_api.all_tasks()

    verify("\n".join([str(task) for task in tasks]), encoding="utf-8")
