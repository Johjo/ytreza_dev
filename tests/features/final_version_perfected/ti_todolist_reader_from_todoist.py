import os

from approvaltests import verify # type: ignore
from dotenv import load_dotenv

from ytreza_dev.features.start_fvp_use_case.todolist_reader_from_todoist import TodolistReaderFromTodoist


def test_all_tasks() -> None:
    load_dotenv(dotenv_path=".env.test")
    api_token = os.getenv("TODOIST_API_TOKEN")
    assert api_token, "TODOIST_API_TOKEN not found in .env.test"

    todoist_api = TodolistReaderFromTodoist()
    tasks = todoist_api.all_tasks()

    verify("\n".join([str(task) for task in tasks]), encoding="utf-8")
