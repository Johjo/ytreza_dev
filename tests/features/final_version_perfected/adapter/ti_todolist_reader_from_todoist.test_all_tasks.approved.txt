ExternalTask(name='T√©l√©chargez des [applications et plugins gratuits suppl√©mentaires](https://todoist.com/fr/downloads?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=inbox) pour üíª,‚åöÔ∏è,üñ•,üì± et üìß', url='https://app.todoist.com/app/task/9388991121', id='9388991121')
ExternalTask(name='Connecter mon calendrier professionnel √† Todoist', url='https://app.todoist.com/app/task/9388991491', id='9388991491')
ExternalTask(name='R√©pondez au [questionnaire sur votre m√©thode de productivit√©](https://todoist.com/productivity-methods?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=inbox)', url='https://app.todoist.com/app/task/9388991122', id='9388991122')
ExternalTask(name='Passer en revue ma journ√©e et planifier', url='https://app.todoist.com/app/task/9388991492', id='9388991492')
ExternalTask(name='Ajouter mes routines _professionnelles_', url='https://app.todoist.com/app/task/9388991493', id='9388991493')
ExternalTask(name='[Faire une pause r√©guli√®re](https://blog.doist.com/end-work-day/?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work) üåñ', url='https://app.todoist.com/app/task/9388991494', id='9388991494')
ExternalTask(name='[Les int√©grations de Todoist qui boosteront votre productivit√©](https://blog.doist.com/todoist-integrations-productivity/?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work)', url='https://app.todoist.com/app/task/9388991495', id='9388991495')
ExternalTask(name='Explorer les [mod√®les professionnels de Todoist](https://todoist.com/fr/templates/category/work?utm_source=todoist&utm_medium=in_app&utm_campaign=onboarding_project&utm_content=work)', url='https://app.todoist.com/app/task/9388991496', id='9388991496')
